with-simulated-input.el

```diff
@@ -163,101 +175,24 @@ are propagated normally.
-           (list 'switch-to-buffer correct-current-buffer)
-           ;; Then we run the body form
-           body-form)
-          ;; Then we run each of the actions specified in KEYS
-          (cl-loop
-           for action in keylist
-           if (not (stringp action))
-           collect action)
-          ;; Finally we throw the canary if we read past the end of
-          ;; the input.
-          (list end-of-actions-form)))
-        ;; Wrap each action in a lexical closure so it can refer to
-        ;; variables from the caller.
-        (action-closures
-         (cl-loop
-          for action in action-list
-          collect (wsi-make-closure action lexenv)))
-        ;; Replace non-strings with `next-action-key' and concat
-        ;; everything together
-        (full-key-sequence
-         (cl-loop
-          for action in keylist
-          if (stringp action)
-          collect action into key-sequence-list
-          else
-          collect next-action-key into key-sequence-list
-          finally return
-          ;; Prepend and append `next-action-key' as appropriate to
-          ;; switch buffer, run body, and throw canary.
-          (concat
-           ;; Switch to correct buffer
-           next-action-key " "
-           ;; Start executing body
-           next-action-key " "
-           ;; Execute the actual key sequence
-           (mapconcat #'identity key-sequence-list " ")
-           ;; Throw the canary if BODY reads past the provided input
-           " " next-action-key)))
-        ;; Define the next action command with lexical scope so it can
-        ;; access `action-closures'.
-        ((symbol-function 'wsi-run-next-action)
-         (lambda ()
-           (interactive)
-           (condition-case err
-               (if action-closures
-                   (let ((next-action (pop action-closures)))
-                     (funcall next-action))
-                 (error "`with-simulated-input' reached end of action list without returning"))
-             (error (throw 'wsi-threw-error err)))))
-        ;; Set up the temporary keymap
-        (action-map (make-sparse-keymap)))
-     ;; Finish setting up the keymap for the temp command
-     (define-key action-map (kbd next-action-key) 'wsi-run-next-action)
-     (setq
-      thrown-error
-      (catch 'wsi-threw-error
-        (setq
-         result
-         (catch 'wsi-body-finished
-           (let ((overriding-terminal-local-map action-map))
-             (execute-kbd-macro (kbd full-key-sequence)))))
-        ;; If we got here, then no error
-        (throw 'wsi-threw-error nil)))
-     (when thrown-error
-       (signal (car thrown-error) (cdr thrown-error)))
-     (if (eq result wsi--canary-sym)
-         (error "Reached end of simulated input while evaluating body")
-       result)))
+  (declare (indent 1) (debug ([&or ("quote" (&rest &or stringp def-form))
+                                   (&rest &or stringp def-form)
```

Comment:
The Elisp manual says that `def-form` "may appear only after `&define`". Does that make it invalid here?

Reply:
You are right, fixed. The Elisp manual confusingly says that if a `body` is wrapped in a lambda, one should use `def-form`, even though the latter is to be used only for `&define`.